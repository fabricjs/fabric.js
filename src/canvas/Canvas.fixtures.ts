import { version } from '../../package.json';

export const EMPTY_JSON = { version: version, objects: [] };

export const PATH_JSON = {
  version: version,
  objects: [
    {
      version: version,
      type: 'path',
      originX: 'left',
      originY: 'top',
      left: 268,
      top: 266,
      width: 51,
      height: 49,
      fill: 'rgb(0,0,0)',
      stroke: null,
      strokeWidth: 1,
      scaleX: 1,
      scaleY: 1,
      angle: 0,
      flipX: false,
      flipY: false,
      opacity: 1,
      path: [
        ['M', 18.511, 13.99],
        ['c', 0, 0, -2.269, -4.487, -12.643, 4.411],
        ['c', 0, 0, 4.824, -14.161, 19.222, -9.059],
        ['l', 0.379, -2.1],
        ['c', -0.759, -0.405, -1.375, -1.139, -1.645, -2.117],
        ['c', -0.531, -1.864, 0.371, -3.854, 1.999, -4.453],
        ['c', 0.312, -0.118, 0.633, -0.169, 0.953, -0.169],
        ['c', 1.299, 0, 2.514, 0.953, 2.936, 2.455],
        ['c', 0.522, 1.864, -0.372, 3.854, -1.999, 4.453],
        ['c', -0.229, 0.084, -0.464, 0.127, -0.692, 0.152],
        ['l', -0.379, 2.37],
        ['c', 1.146, 0.625, 2.024, 1.569, 2.674, 2.758],
        ['c', 3.213, 2.514, 8.561, 4.184, 11.774, -8.232],
        ['c', 0, 0, 0.86, 16.059, -12.424, 14.533],
        ['c', 0.008, 2.859, 0.615, 5.364, -0.076, 8.224],
        ['c', 8.679, 3.146, 15.376, 14.389, 17.897, 18.168],
        ['l', 2.497, -2.151],
        ['l', 1.206, 1.839],
        ['l', -3.889, 3.458],
        ['C', 46.286, 48.503, 31.036, 32.225, 22.72, 35.81],
        ['c', -1.307, 2.851, -3.56, 6.891, -7.481, 8.848],
        ['c', -4.689, 2.336, -9.084, -0.802, -11.277, -2.868],
        ['l', -1.948, 3.104],
        ['l', -1.628, -1.333],
        ['l', 3.138, -4.689],
        ['c', 0.025, 0, 9, 1.932, 9, 1.932],
        ['c', 0.877, -9.979, 2.893, -12.905, 4.942, -15.621],
        ['C', 17.878, 21.775, 18.713, 17.397, 18.511, 13.99],
        ['z', null],
      ],
    },
  ],
  background: '#ff5555',
  overlay: 'rgba(0,0,0,0.2)',
};

export const PATH_WITHOUT_DEFAULTS_JSON = {
  version: version,
  objects: [
    {
      version: version,
      type: 'path',
      left: 268,
      top: 266,
      width: 51,
      height: 49,
      path: [
        ['M', 18.511, 13.99],
        ['c', 0, 0, -2.269, -4.487, -12.643, 4.411],
        ['c', 0, 0, 4.824, -14.161, 19.222, -9.059],
        ['l', 0.379, -2.1],
        ['c', -0.759, -0.405, -1.375, -1.139, -1.645, -2.117],
        ['c', -0.531, -1.864, 0.371, -3.854, 1.999, -4.453],
        ['c', 0.312, -0.118, 0.633, -0.169, 0.953, -0.169],
        ['c', 1.299, 0, 2.514, 0.953, 2.936, 2.455],
        ['c', 0.522, 1.864, -0.372, 3.854, -1.999, 4.453],
        ['c', -0.229, 0.084, -0.464, 0.127, -0.692, 0.152],
        ['l', -0.379, 2.37],
        ['c', 1.146, 0.625, 2.024, 1.569, 2.674, 2.758],
        ['c', 3.213, 2.514, 8.561, 4.184, 11.774, -8.232],
        ['c', 0, 0, 0.86, 16.059, -12.424, 14.533],
        ['c', 0.008, 2.859, 0.615, 5.364, -0.076, 8.224],
        ['c', 8.679, 3.146, 15.376, 14.389, 17.897, 18.168],
        ['l', 2.497, -2.151],
        ['l', 1.206, 1.839],
        ['l', -3.889, 3.458],
        ['C', 46.286, 48.503, 31.036, 32.225, 22.72, 35.81],
        ['c', -1.307, 2.851, -3.56, 6.891, -7.481, 8.848],
        ['c', -4.689, 2.336, -9.084, -0.802, -11.277, -2.868],
        ['l', -1.948, 3.104],
        ['l', -1.628, -1.333],
        ['l', 3.138, -4.689],
        ['c', 0.025, 0, 9, 1.932, 9, 1.932],
        ['c', 0.877, -9.979, 2.893, -12.905, 4.942, -15.621],
        ['C', 17.878, 21.775, 18.713, 17.397, 18.511, 13.99],
        ['z', null],
      ],
    },
  ],
  background: '#ff5555',
  overlay: 'rgba(0,0,0,0.2)',
};

export const PATH_OBJ_JSON = {
  version: version,
  type: 'path',
  originX: 'left',
  originY: 'top',
  left: 268,
  top: 266,
  width: 51,
  height: 49,
  fill: 'rgb(0,0,0)',
  stroke: null,
  strokeWidth: 1,
  scaleX: 1,
  scaleY: 1,
  angle: 0,
  flipX: false,
  flipY: false,
  opacity: 1,
  path: [
    ['M', 18.511, 13.99],
    ['c', 0, 0, -2.269, -4.487, -12.643, 4.411],
    ['c', 0, 0, 4.824, -14.161, 19.222, -9.059],
    ['l', 0.379, -2.1],
    ['c', -0.759, -0.405, -1.375, -1.139, -1.645, -2.117],
    ['c', -0.531, -1.864, 0.371, -3.854, 1.999, -4.453],
    ['c', 0.312, -0.118, 0.633, -0.169, 0.953, -0.169],
    ['c', 1.299, 0, 2.514, 0.953, 2.936, 2.455],
    ['c', 0.522, 1.864, -0.372, 3.854, -1.999, 4.453],
    ['c', -0.229, 0.084, -0.464, 0.127, -0.692, 0.152],
    ['l', -0.379, 2.37],
    ['c', 1.146, 0.625, 2.024, 1.569, 2.674, 2.758],
    ['c', 3.213, 2.514, 8.561, 4.184, 11.774, -8.232],
    ['c', 0, 0, 0.86, 16.059, -12.424, 14.533],
    ['c', 0.008, 2.859, 0.615, 5.364, -0.076, 8.224],
    ['c', 8.679, 3.146, 15.376, 14.389, 17.897, 18.168],
    ['l', 2.497, -2.151],
    ['l', 1.206, 1.839],
    ['l', -3.889, 3.458],
    ['C', 46.286, 48.503, 31.036, 32.225, 22.72, 35.81],
    ['c', -1.307, 2.851, -3.56, 6.891, -7.481, 8.848],
    ['c', -4.689, 2.336, -9.084, -0.802, -11.277, -2.868],
    ['l', -1.948, 3.104],
    ['l', -1.628, -1.333],
    ['l', 3.138, -4.689],
    ['c', 0.025, 0, 9, 1.932, 9, 1.932],
    ['c', 0.877, -9.979, 2.893, -12.905, 4.942, -15.621],
    ['C', 17.878, 21.775, 18.713, 17.397, 18.511, 13.99],
    ['z', null],
  ],
};

export const PATH_DATALESS_JSON = {
  version: version,
  objects: [
    {
      type: 'Path',
      version: version,
      originX: 'center',
      originY: 'center',
      left: 200,
      top: 200,
      width: 200,
      height: 200,
      fill: 'rgb(0,0,0)',
      stroke: null,
      strokeWidth: 1,
      strokeDashArray: null,
      strokeLineCap: 'butt',
      strokeDashOffset: 0,
      strokeLineJoin: 'miter',
      strokeUniform: false,
      strokeMiterLimit: 4,
      scaleX: 1,
      scaleY: 1,
      angle: 0,
      flipX: false,
      flipY: false,
      opacity: 1,
      shadow: null,
      visible: true,
      backgroundColor: '',
      fillRule: 'nonzero',
      paintFirst: 'fill',
      globalCompositeOperation: 'source-over',
      skewX: 0,
      skewY: 0,
      sourcePath: 'http://example.com/',
    },
  ],
};

export const RECT_JSON = {
  version: version,
  objects: [
    {
      type: 'Rect',
      version: version,
      originX: 'center',
      originY: 'center',
      left: 0,
      top: 0,
      width: 10,
      height: 10,
      fill: 'rgb(0,0,0)',
      stroke: null,
      strokeWidth: 1,
      strokeDashArray: null,
      strokeLineCap: 'butt',
      strokeDashOffset: 0,
      strokeLineJoin: 'miter',
      strokeUniform: false,
      strokeMiterLimit: 4,
      scaleX: 1,
      scaleY: 1,
      angle: 0,
      flipX: false,
      flipY: false,
      opacity: 1,
      shadow: null,
      visible: true,
      backgroundColor: '',
      fillRule: 'nonzero',
      paintFirst: 'fill',
      globalCompositeOperation: 'source-over',
      skewX: 0,
      skewY: 0,
      rx: 0,
      ry: 0,
    },
  ],
  background: '#ff5555',
  overlay: 'rgba(0,0,0,0.2)',
};
