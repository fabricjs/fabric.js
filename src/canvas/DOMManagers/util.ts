import { LEFT, NONE, TOP } from '../../constants';
import type { TSize } from '../../typedefs';
import {
  getDocumentFromElement,
  getWindowFromElement,
  getScrollLeftTop,
} from '../../util/dom_misc';
import { setStyle } from '../../util/dom_style';

export const setCanvasDimensions = (
  el: HTMLCanvasElement,
  ctx: CanvasRenderingContext2D,
  { width, height }: TSize,
  retinaScaling = 1
) => {
  el.width = width;
  el.height = height;
  if (retinaScaling > 1) {
    el.setAttribute('width', (width * retinaScaling).toString());
    el.setAttribute('height', (height * retinaScaling).toString());
    ctx.scale(retinaScaling, retinaScaling);
  }
};

export function allowTouchScrolling(element: HTMLElement, allow: boolean) {
  const touchAction = allow ? 'manipulation' : NONE;
  setStyle(element, {
    'touch-action': touchAction,
    '-ms-touch-action': touchAction,
  });
}

export type CSSDimensions = {
  width: number | `${number}${string}`;
  height: number | `${number}${string}`;
};

export const setCSSDimensions = (
  el: HTMLElement,
  { width, height }: Partial<CSSDimensions>
) => {
  width && (el.style.width = typeof width === 'number' ? `${width}px` : width);
  height &&
    (el.style.height = typeof height === 'number' ? `${height}px` : height);
};

/**
 * Returns offset for a given element
 * @param {HTMLElement} element Element to get offset for
 * @return {Object} Object with "left" and "top" properties
 */
export function getElementOffset(element: HTMLElement) {
  let box = { left: 0, top: 0 };
  const doc = element && getDocumentFromElement(element),
    offset = { left: 0, top: 0 },
    offsetAttributes = {
      borderLeftWidth: LEFT,
      borderTopWidth: TOP,
      paddingLeft: LEFT,
      paddingTop: TOP,
    } as const;

  if (!doc) {
    return offset;
  }
  const elemStyle =
    getWindowFromElement(element)?.getComputedStyle(element, null) || {};
  for (const attr in offsetAttributes) {
    // @ts-expect-error TS learn to iterate!
    offset[offsetAttributes[attr]] += parseInt(elemStyle[attr], 10) || 0;
  }

  const docElem = doc.documentElement;
  if (typeof element.getBoundingClientRect !== 'undefined') {
    box = element.getBoundingClientRect();
  }

  const scrollLeftTop = getScrollLeftTop(element);

  return {
    left:
      box.left + scrollLeftTop.left - (docElem.clientLeft || 0) + offset.left,
    top: box.top + scrollLeftTop.top - (docElem.clientTop || 0) + offset.top,
  };
}

/**
 * Makes element unselectable
 * @param {HTMLElement} element Element to make unselectable
 * @return {HTMLElement} Element that was passed in
 */
export function makeElementUnselectable(element: HTMLElement) {
  if (typeof element.onselectstart !== 'undefined') {
    element.onselectstart = () => false;
  }
  element.style.userSelect = NONE;
  return element;
}
