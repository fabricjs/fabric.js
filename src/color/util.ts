/**
 * @param {Number} p
 * @param {Number} q
 * @param {Number} t
 * @return {Number}
 */
export function hue2rgb(p: number, q: number, t: number): number {
  if (t < 0) {
    t += 1;
  }
  if (t > 1) {
    t -= 1;
  }
  if (t < 1 / 6) {
    return p + (q - p) * 6 * t;
  }
  if (t < 1 / 2) {
    return q;
  }
  if (t < 2 / 3) {
    return p + (q - p) * (2 / 3 - t) * 6;
  }
  return p;
}

/**
 * Convert a value âˆˆ [0, 255] to hex
 */
export function hexify(value: number) {
  const hexValue = value.toString(16).toUpperCase();
  return hexValue.length === 1 ? `0${hexValue}` : hexValue;
}
