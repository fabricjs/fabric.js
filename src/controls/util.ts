import {
  TPointerEvent,
  Transform,
  TransformAction,
  BasicTransformEvent,
} from '../EventTypeDefs';
import { resolveOrigin } from '../util/misc/resolveOrigin';
import { Point } from '../Point';
import type { FabricObject } from '../shapes/Object/FabricObject';
import { TOriginX, TOriginY } from '../typedefs';
import {
  degreesToRadians,
  radiansToDegrees,
} from '../util/misc/radiansDegreesConversion';
import type { Control } from './Control';

export const NOT_ALLOWED_CURSOR = 'not-allowed';

/**
 * @param {Boolean} alreadySelected true if target is already selected
 * @param {String} corner a string representing the corner ml, mr, tl ...
 * @param {Event} e Event object
 * @param {FabricObject} [target] inserted back to help overriding. Unused
 */
export const getActionFromCorner = (
  alreadySelected: boolean,
  corner: string,
  e: TPointerEvent,
  target: FabricObject
) => {
  if (!corner || !alreadySelected) {
    return 'drag';
  }
  const control = target.controls[corner];
  return control.getActionName(e, control, target);
};

/**
 * Checks if transform is centered
 * @param {Object} transform transform data
 * @return {Boolean} true if transform is centered
 */
export function isTransformCentered(transform: Transform) {
  return transform.originX === 'center' && transform.originY === 'center';
}

export function invertOrigin(origin: TOriginX | TOriginY) {
  return -resolveOrigin(origin) + 0.5;
}

export const isLocked = (
  target: FabricObject,
  lockingKey:
    | 'lockMovementX'
    | 'lockMovementY'
    | 'lockRotation'
    | 'lockScalingX'
    | 'lockScalingY'
    | 'lockSkewingX'
    | 'lockSkewingY'
    | 'lockScalingFlip'
) => target[lockingKey];

export const commonEventInfo: TransformAction<
  Transform,
  BasicTransformEvent
> = (eventData, transform, x, y) => {
  return {
    e: eventData,
    transform,
    pointer: new Point(x, y),
  };
};

/**
 * Combine control position and object angle to find the control direction compared
 * to the object center.
 * @param {FabricObject} fabricObject the fabric object for which we are rendering controls
 * @param {Control} control the control class
 * @return {Number} 0 - 7 a quadrant number
 */
export function findCornerQuadrant(
  fabricObject: FabricObject,
  control: Control
) {
  //  angle is relative to canvas plane
  const angle = fabricObject.getTotalAngle(),
    cornerAngle =
      angle + radiansToDegrees(Math.atan2(control.y, control.x)) + 360;
  return Math.round((cornerAngle % 360) / 45);
}

/**
 * @returns the normalized point (rotated relative to center) in local coordinates
 */
function normalizePoint(
  target: FabricObject,
  point: Point,
  originX: TOriginX,
  originY: TOriginY
): Point {
  const center = target.getRelativeCenterPoint(),
    p =
      typeof originX !== 'undefined' && typeof originY !== 'undefined'
        ? target.translateToGivenOrigin(
            center,
            'center',
            'center',
            originX,
            originY
          )
        : new Point(target.left, target.top),
    p2 = target.angle
      ? point.rotate(-degreesToRadians(target.angle), center)
      : point;
  return p2.subtract(p);
}

/**
 * Transforms a point to the offset from the given origin
 * @param {Object} transform
 * @param {String} originX
 * @param {String} originY
 * @param {number} x
 * @param {number} y
 * @return {Fabric.Point} the normalized point
 */
export function getLocalPoint(
  { target, corner }: Transform,
  originX: TOriginX,
  originY: TOriginY,
  x: number,
  y: number
) {
  const control = target.controls[corner],
    zoom = target.canvas?.getZoom() || 1,
    padding = target.padding / zoom,
    localPoint = normalizePoint(target, new Point(x, y), originX, originY);
  if (localPoint.x >= padding) {
    localPoint.x -= padding;
  }
  if (localPoint.x <= -padding) {
    localPoint.x += padding;
  }
  if (localPoint.y >= padding) {
    localPoint.y -= padding;
  }
  if (localPoint.y <= padding) {
    localPoint.y += padding;
  }
  localPoint.x -= control.offsetX;
  localPoint.y -= control.offsetY;
  return localPoint;
}
