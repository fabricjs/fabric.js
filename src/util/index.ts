export { cos } from './misc/cos';
export { sin } from './misc/sin';
export {
  rotateVector,
  createVector,
  calcAngleBetweenVectors,
  getUnitVector,
  calcVectorRotation,
  crossProduct,
  dotProduct,
  getOrthonormalVector,
  isBetweenVectors,
  magnitude,
} from './misc/vectors';
export {
  degreesToRadians,
  radiansToDegrees,
} from './misc/radiansDegreesConversion';
export { rotatePoint } from './misc/rotatePoint';
export * from './misc/projectStroke';
export type {
  TComposeMatrixArgs,
  TQrDecomposeOut,
  TRotateMatrixArgs,
  TScaleMatrixArgs,
  TTranslateMatrixArgs,
} from './misc/matrix';
export {
  transformPoint,
  invertTransform,
  composeMatrix,
  qrDecompose,
  createTranslateMatrix,
  createRotateMatrix,
  createScaleMatrix,
  createSkewXMatrix,
  createSkewYMatrix,
  calcDimensionsMatrix,
  multiplyTransformMatrices,
  multiplyTransformMatrixArray,
  isIdentityMatrix,
} from './misc/matrix';
export type { TextStyleArray } from './misc/textStyles';
export {
  stylesFromArray,
  stylesToArray,
  hasStyleChanged,
} from './misc/textStyles';
export {
  createCanvasElement,
  createImage,
  copyCanvasElement,
  toDataURL,
} from './misc/dom';
export { toFixed } from './misc/toFixed';
export {
  matrixToSVG,
  parsePreserveAspectRatioAttribute,
  parseUnit,
  getSvgAttributes,
} from './misc/svgParsing';
export { groupSVGElements } from './misc/groupSVGElements';
export { findScaleToFit, findScaleToCover } from './misc/findScaleTo';
export { capValue } from './misc/capValue';
export {
  saveObjectTransform,
  resetObjectTransform,
  addTransformToObject,
  applyTransformToObject,
  removeTransformFromObject,
  sizeAfterTransform,
} from './misc/objectTransforms';
export { makeBoundingBoxFromPoints } from './misc/boundingBoxFromPoints';
export {
  calcPlaneChangeMatrix,
  sendPointToPlane,
  sendVectorToPlane,
  sendObjectToPlane,
} from './misc/planeChange';
export * as string from './lang_string';
export type {
  EnlivenObjectOptions,
  LoadImageOptions,
} from './misc/objectEnlive';
export {
  loadImage,
  enlivenObjects,
  enlivenObjectEnlivables,
} from './misc/objectEnlive';
export { pick } from './misc/pick';
export * from './path/typedefs';
export {
  joinPath,
  parsePath,
  makePathSimpler,
  getSmoothPathFromPoints,
  getPathSegmentsInfo,
  getBoundsOfCurve,
  getPointOnPath,
  transformPath,
  getRegularPolygonPath,
} from './path';
export { setStyle } from './dom_style';
export { isTouchEvent, getPointer } from './dom_event';
export { isTransparent } from './misc/isTransparent';
export { mergeClipPaths } from './misc/mergeClipPaths';
export * from './animation';
export * as ease from './animation/easing';
export {
  requestAnimFrame,
  cancelAnimFrame,
} from './animation/AnimationFrameProvider';
export { removeFromArray } from './internals/removeFromArray';
export { getRandomInt } from './internals/getRandomInt';
export { request } from './internals/dom_request';

// for test compatibility. We don't want to export it
export { removeTransformMatrixForSvgParsing } from './transform_matrix_removal';
