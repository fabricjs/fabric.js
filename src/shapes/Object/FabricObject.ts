import { ObjectEvents } from '../../EventTypeDefs';
import { FabricObjectSVGExportMixin } from './FabricObjectSVGExportMixin';
import { InteractiveFabricObject } from './InteractiveObject';
import { applyMixins } from '../../util/applyMixins';
import { StatefulMixin } from '../../mixins/stateful.mixin';

// TODO somehow we have to make a tree-shakeable import

// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars
export interface FabricObject<EventSpec extends ObjectEvents = ObjectEvents>
  extends FabricObjectSVGExportMixin {}

export class FabricObject<
  EventSpec extends ObjectEvents = ObjectEvents
> extends InteractiveFabricObject<EventSpec> {}

applyMixins(FabricObject, [FabricObjectSVGExportMixin, StatefulMixin]);

export { cacheProperties, stateProperties } from './defaultValues';
