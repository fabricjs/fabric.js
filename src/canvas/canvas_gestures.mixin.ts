//@ts-nocheck

import { scalingEqually } from '../controls/scale';
import { fireEvent } from '../util/fireEvent';
import {
  degreesToRadians,
  radiansToDegrees,
} from '../util/misc/radiansDegreesConversion';
import { Canvas } from './Canvas';

/**
 * Adds support for multi-touch gestures using the Event.js library.
 * Fires the following custom events:
 * - touch:gesture
 * - touch:drag
 * - touch:orientation
 * - touch:shake
 * - touch:longpress
 */
Object.assign(Canvas.prototype, {
  /**
   * Method that defines actions when an Event.js gesture is detected on an object. Currently only supports
   * 2 finger gestures.
   * @param {Event} e Event object by Event.js
   * @param {Event} self Event proxy object by Event.js
   */
  __onTransformGesture: function (e, self) {
    if (
      this.isDrawingMode ||
      !e.touches ||
      e.touches.length !== 2 ||
      'gesture' !== self.gesture
    ) {
      return;
    }

    var target = this.findTarget(e);
    if ('undefined' !== typeof target) {
      this.__gesturesParams = {
        e: e,
        self: self,
        target: target,
      };

      this.__gesturesRenderer();
    }

    this.fire('touch:gesture', {
      target: target,
      e: e,
      self: self,
    });
  },
  __gesturesParams: null,
  __gesturesRenderer: function () {
    if (this.__gesturesParams === null || this._currentTransform === null) {
      return;
    }

    var self = this.__gesturesParams.self,
      t = this._currentTransform,
      e = this.__gesturesParams.e;

    t.action = 'scale';
    t.originX = t.originY = 'center';

    this._scaleObjectBy(self.scale, e);

    if (self.rotation !== 0) {
      t.action = 'rotate';
      this._rotateObjectByAngle(self.rotation, e);
    }

    this.requestRenderAll();

    t.action = 'drag';
  },

  /**
   * Method that defines actions when an Event.js drag is detected.
   *
   * @param {Event} e Event object by Event.js
   * @param {Event} self Event proxy object by Event.js
   */
  __onDrag: function (e, self) {
    this.fire('touch:drag', {
      e: e,
      self: self,
    });
  },

  /**
   * Method that defines actions when an Event.js orientation event is detected.
   *
   * @param {Event} e Event object by Event.js
   * @param {Event} self Event proxy object by Event.js
   */
  __onOrientationChange: function (e, self) {
    this.fire('touch:orientation', {
      e: e,
      self: self,
    });
  },

  /**
   * Method that defines actions when an Event.js shake event is detected.
   *
   * @param {Event} e Event object by Event.js
   * @param {Event} self Event proxy object by Event.js
   */
  __onShake: function (e, self) {
    this.fire('touch:shake', {
      e: e,
      self: self,
    });
  },

  /**
   * Method that defines actions when an Event.js longpress event is detected.
   *
   * @param {Event} e Event object by Event.js
   * @param {Event} self Event proxy object by Event.js
   */
  __onLongPress: function (e, self) {
    this.fire('touch:longpress', {
      e: e,
      self: self,
    });
  },

  /**
   * @private
   * @param {Event} [e] Event object fired on Event.js gesture
   * @param {Event} [self] Inner Event object
   */
  _onGesture: function (e, self) {
    this.__onTransformGesture(e, self);
  },

  /**
   * @private
   * @param {Event} [e] Event object fired on Event.js drag
   * @param {Event} [self] Inner Event object
   */
  _onDrag: function (e, self) {
    this.__onDrag(e, self);
  },

  /**
   * Scales an object by a factor
   * @param {Number} s The scale factor to apply to the current scale level
   * @param {Event} e Event object by Event.js
   */
  _scaleObjectBy: function (s, e) {
    var t = this._currentTransform,
      target = t.target;
    t.gestureScale = s;
    target._scaling = true;
    return scalingEqually(e, t, 0, 0);
  },

  /**
   * Rotates object by an angle
   * @param {Number} curAngle The angle of rotation in degrees
   * @param {Event} e Event object by Event.js
   */
  _rotateObjectByAngle: function (curAngle, e) {
    var t = this._currentTransform;

    if (t.target.get('lockRotation')) {
      return;
    }
    t.target.rotate(radiansToDegrees(degreesToRadians(curAngle) + t.theta));
    fireEvent('rotating', {
      target: t.target,
      e: e,
      transform: t,
    });
  },

  /**
   * @private
   * @param {Event} [e] Event object fired on Event.js orientation change
   * @param {Event} [self] Inner Event object
   */
  _onOrientationChange: function (e, self) {
    this.__onOrientationChange(e, self);
  },

  /**
   * @private
   * @param {Event} [e] Event object fired on Event.js shake
   * @param {Event} [self] Inner Event object
   */
  _onShake: function (e, self) {
    this.__onShake(e, self);
  },

  /**
   * @private
   * @param {Event} [e] Event object fired on Event.js shake
   * @param {Event} [self] Inner Event object
   */
  _onLongPress: function (e, self) {
    this.__onLongPress(e, self);
  },
});
