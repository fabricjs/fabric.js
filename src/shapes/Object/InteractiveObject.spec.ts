import { radiansToDegrees } from '../../util';
import { Group } from '../Group';
import { Canvas } from '../../canvas/Canvas';
import { FabricObject } from './FabricObject';
import type { TOCoord } from './InteractiveObject';

describe('FabricObject', () => {
  it('Interactive + BaseObject default values', () => {
    const { controls, ...defaults } = FabricObject.getDefaults();
    expect(defaults).toMatchSnapshot();
  });
  describe('setCoords for objects inside group with rotation', () => {
    it('all corners are rotated as much as the object total angle', () => {
      const canvas = new Canvas();
      const object = new FabricObject({
        left: 25,
        top: 60,
        width: 75,
        height: 100,
        angle: 10,
        scaleY: 2,
        fill: 'blue',
      });
      const group = new Group([object], {
        angle: 30,
        scaleX: 2,
        interactive: true,
        subTargetCheck: true,
      });
      canvas.add(group);
      group.setCoords();
      const objectAngle = Math.round(object.getTotalAngle());
      expect(objectAngle).toEqual(35);
      Object.values(object.oCoords).forEach((cornerPoint: TOCoord) => {
        const controlAngle = Math.round(
          radiansToDegrees(
            Math.atan2(
              cornerPoint.corner.tr.y - cornerPoint.corner.tl.y,
              cornerPoint.corner.tr.x - cornerPoint.corner.tl.x
            )
          )
        );
        expect(controlAngle).toEqual(objectAngle);
      });
    });
  });
});
