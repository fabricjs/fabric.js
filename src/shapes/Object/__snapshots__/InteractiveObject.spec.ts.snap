// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FabricObject Interactive + BaseObject default values 1`] = `
{
  "FX_DURATION": 500,
  "absolutePositioned": false,
  "activeOn": "down",
  "angle": 0,
  "backgroundColor": "",
  "borderColor": "rgb(178,204,255)",
  "borderDashArray": null,
  "borderOpacityWhenMoving": 0.4,
  "borderScaleFactor": 1,
  "centeredRotation": true,
  "centeredScaling": false,
  "clipPath": undefined,
  "colorProperties": [
    "fill",
    "stroke",
    "backgroundColor",
  ],
  "cornerColor": "rgb(178,204,255)",
  "cornerDashArray": null,
  "cornerSize": 13,
  "cornerStrokeColor": "",
  "cornerStyle": "rect",
  "dirty": true,
  "evented": true,
  "excludeFromExport": false,
  "fill": "rgb(0,0,0)",
  "fillRule": "nonzero",
  "flipX": false,
  "flipY": false,
  "globalCompositeOperation": "source-over",
  "hasBorders": true,
  "hasControls": true,
  "height": 0,
  "hoverCursor": null,
  "includeDefaultValues": true,
  "inverted": false,
  "left": 0,
  "lockMovementX": false,
  "lockMovementY": false,
  "lockRotation": false,
  "lockScalingFlip": false,
  "lockScalingX": false,
  "lockScalingY": false,
  "lockSkewingX": false,
  "lockSkewingY": false,
  "minScaleLimit": 0,
  "moveCursor": null,
  "noScaleCache": true,
  "objectCaching": true,
  "opacity": 1,
  "originX": "left",
  "originY": "top",
  "padding": 0,
  "paintFirst": "fill",
  "perPixelTargetFind": false,
  "scaleX": 1,
  "scaleY": 1,
  "selectable": true,
  "selectionBackgroundColor": "",
  "shadow": null,
  "skewX": 0,
  "skewY": 0,
  "stroke": null,
  "strokeDashArray": null,
  "strokeDashOffset": 0,
  "strokeLineCap": "butt",
  "strokeLineJoin": "miter",
  "strokeMiterLimit": 4,
  "strokeUniform": false,
  "strokeWidth": 1,
  "top": 0,
  "touchCornerSize": 24,
  "transparentCorners": true,
  "visible": true,
  "width": 0,
}
`;
